pyquil/compatibility/v2/api/_qam.py

```diff
@@ -0,0 +1,51 @@
+from typing import Optional, Sequence, Union
+
+from rpcq.messages import ParameterAref
+from pyquil.api._qam import QAM, QAMExecutionResult, QuantumExecutable
+
+
+class StatefulQAM(QAM):
+    _loaded_executable: Optional[QuantumExecutable]
```

Comment:
I still think we should avoid using class variables as if they were instance variables. Every time I've tried to search on this, I come up with guides like [this](https://www.digitalocean.com/community/tutorials/understanding-class-and-instance-variables-in-python-3#working-with-class-and-instance-variables-together) and [this](https://stackoverflow.com/a/9056994).

We're not setting a default value that is something mutable like a dict, so it doesn't matter much in this specific case. But I still think we should avoid it, as it relies on a quirk of python in that you can override/shadow a class variable with an instance variable by setting it on `self`.

That's my argument, but it's your call. I know there are some editor-related downsides to doing it the recommended way.

Reply:
I agree, but this is not a class variable. It's never set; it's only an annotation. Python's own docs say that this is the correct way to type instance attributes. There's [ClassVar](https://docs.python.org/3/library/typing.html#typing.ClassVar) to indicate class variables (note the example there of annotating an instance variable). Given the choice, I'd actually prefer these all to be restored where they're now missing on `rc`.

Reply:
I see your logic about not setting it. I think this is always going to be a place where I think python chooses to be confusing. Willing to do what you want here.

What is confusing is that the doc you linked does have the value set and still calls the thing an "instance variable", when it's definitely not. I will say that there's a reason they added the `ClassVar` to the `int` and not the `dict` in their example, because you get funniness if you had:

```python
from typing import ClassVar, Dict


class Starship:
    stats: Dict[str, int] = {}
    damage: ClassVar[int] = 10


s1 = Starship()
s2 = Starship()

print(s1.stats)
print(s2.stats)
s1.stats["xp"] = 42
print(s1.stats)
print(s2.stats)
```

you'd get:

```
{}
{}
{'xp': 42}
{'xp': 42}
```
(Really, you'd have this funniness regardless of where the `ClassVar` was because it has no runtime behavior.)

If the default value is mutable, the thing suddenly reveals itself as a class variable all along, until you override it by calling `self.stats = {...}` internally or `s1.stats = {...}` externally, which "turns" it into an instance variable (but what if your runtime code never does this?). The fact that you can treat things like instance members in only specific cases with this notation is what causes it to be confusing and inconsistent.

All that said, I personally think it's setting us up for confusion and bugs (or at best, creates a strange rule to remember: "never set a mutable default value with this notation"), but I'm willing to go along with consensus.

Reply:
> I agree, but this is not a class variable. It's never set; it's only an annotation.

This is true. I guess you could say it's neither a class variable nor an instance variable until it has a value. If that value was set on the instance, it becomes an instance variable, and if it was set on the class (either via a default value or explicitly calling `Starship.some_var = ...`, then it's a class variable.

So to clarify when I said "reveals itself as a class variable all along", I was referring to the doc's case where they set a default value. In your own code here in this PR, I agree that it's not a class variable (not is it an instance variable).

Reply:
Really it comes down to the fact that this notation can encourage the setting of default values this way, where things begin to get weird. As long as we don't do that we're in the clear and it's a nice notation tool. But do we want to have to remember that rule? (rhetorical)

Reply:
> On this note, did you intend for EngagementManager._lock to be a class var? Seems like it only needs to synchronize on the instance

You're right -- that was an accident! I was absent-minded and didn't follow my own rule, lol

Reply:
> I find it adds useful information to see which attributes a class is expected to hold without having to parse through its constructor.
Definitely agree on this. I think if we simply avoid defaults, it's a good compromise. Too bad I didn't do that with `EngagementManager` ;)